overlord/snapstate/handlers.go

```diff
@@ -115,6 +115,88 @@ func snapSetupAndState(t *state.Task) (*SnapSetup, *SnapState, error) {
 
 */
 
+const defaultCoreSnapName = "core"
+const defaultBaseSnapsChannel = "stable"
+
+func (m *SnapManager) doPrerequisites(t *state.Task, _ *tomb.Tomb) error {
+	// the state lock is not enough to ensure this is not run in
+	// parallel as it is dropped in snapstate.Install() below
+	m.prerequisitesLock.Lock()
+	defer m.prerequisitesLock.Unlock()
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// check if core is installed already
+	_, err := CoreInfo(st)
+	if err == nil {
+		return nil
+	}
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// check if we need to inject tasks to install core
+	snapsup, _, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.Name()
+	if snapName == defaultCoreSnapName || snapName == "ubuntu-core" {
+		return nil
+	}
+
+	waitNextHelper := func(chg *state.Change, ts *state.TaskSet) {
+		for _, t := range chg.Tasks() {
+			t.Kind()
+			t.WaitAll(ts)
+		}
+	}
+
+	// check that there is no task that installs core already
+	for _, chg := range st.Changes() {
+		if chg.Status().Ready() || chg.ID() == t.Change().ID() {
+			continue
+		}
+		for _, tc := range chg.Tasks() {
+			if tc.Kind() == "link-snap" {
+				snapsup, err := TaskSnapSetup(tc)
+				if err != nil {
+					return err
+				}
+				// some other change aleady installs core
+				if snapsup.Name() == defaultCoreSnapName {
+					// if something else installs core
+					// we need to wait for it
+					waitNextHelper(t.Change(), state.NewTaskSet(chg.Tasks()...))
+					return nil
+				}
+			}
+		}
+	}
+
+	// not installed, nor queued for install -> install it
+	ts, err := Install(st, defaultCoreSnapName, defaultBaseSnapsChannel, snap.R(0), snapsup.UserID, Flags{})
```

Comment:
we need consider that people can still do "snap install core" and here we release the state lock, so this can return with a conflict I think and then we need to setup waiting instead